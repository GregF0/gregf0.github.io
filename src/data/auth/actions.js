import { firebase } from '../../firebase/firebase'

export const setUserInfo = (userInfo) => ({
  type: 'SET_USER_INFO',
  userInfo,
})

export const linkAccount = (authProvider) => {
  return async (dispatch) => {
    try {
      const res = await firebase.auth().currentUser.linkWithPopup(authProvider)

      const user = res.user

      const userInfo = { uid: user.uid }

      userInfo.creationTime = user.metadata.creationTime

      if (user.providerData && user.providerData.length) {
        userInfo.name = user.providerData[0].displayName
        userInfo.photo = user.providerData[0].photoURL
      }

      dispatch(setUserInfo(userInfo))
      return res
    } catch (e) {
      if (e.code === 'auth/credential-already-in-use') {
        return firebase.auth().signInWithCredential(e.credential)
      }
    }
  }
}

export const signOut = () => ({
  type: 'SIGN_OUT',
})

export const startSignOut = () => {
  return async (dispatch) => {
    await firebase.auth().signOut()
    dispatch(signOut())
  }
}
