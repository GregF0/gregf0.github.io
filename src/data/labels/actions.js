import fs from '../../firebase/firebase'

export const setFullscreenDialog = (fullscreenDialog) => ({
  type: 'SET_FULLSCREEN_DIALOG',
  fullscreenDialog,
})

export const setDesktopDialog = (desktopDialog) => ({
  type: 'SET_DESKTOP_DIALOG',
  desktopDialog,
})

export const setDeleteAlert = (deleteAlert) => ({
  type: 'SET_DELETE_ALERT',
  deleteAlert,
})

export const setMenuOpened = (menuOpened) => ({
  type: 'SET_MENU_OPENED',
  menuOpened,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})

export const setLabelEditting = (labelEditting) => ({
  type: 'SET_LABEL_EDITTING',
  labelEditting,
})

export const setLabelSelected = (labelSelected) => ({
  type: 'SET_LABEL_SELECTED',
  labelSelected,
})

export const addLabel = (label) => ({
  type: 'ADD_LABEL',
  label,
})

export const startAddLabel = (label) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const newLabelRef = fs.collection(`users/${uid}/labels`).doc()

    dispatch(
      addLabel({
        id: newLabelRef.id,
        ...label,
      })
    )
    await newLabelRef.set({ ...label })
  }
}

export const editLabel = (id, updates) => ({
  type: 'EDIT_LABEL',
  id,
  updates,
})

export const startEditLabel = (id, updates) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(editLabel(id, updates))

    await fs.doc(`users/${uid}/labels/${id}`).update({
      ...updates,
    })
  }
}

export const deleteLabel = (id) => ({
  type: 'DELETE_LABEL',
  id,
})

export const startDeleteLabel = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(deleteLabel(id))

    await fs.doc(`users/${uid}/labels/${id}`).delete()
  }
}

export const setLabels = (labels) => ({
  type: 'SET_LABELS',
  labels,
})

export const startSetLabels = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const labelsRef = await fs.collection(`users/${uid}/labels`).get()

      const labels = labelsRef.docs.map((label) => ({
        id: label.id,
        ...label.data(),
      }))

      dispatch(setLabels(labels))

      return labelsRef
    } catch (e) {
      dispatch(setLabels([]))
    }
  }
}
