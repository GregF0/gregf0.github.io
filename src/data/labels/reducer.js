export const initialState = {
  fullscreenDialog: false,
  desktopDialog: false,
  deleteAlert: false,
  menuOpened: false,
  formValue: null,
  labelEditting: null,
  labelSelected: null,
  data: [],
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_FULLSCREEN_DIALOG':
      return {
        ...state,
        fullscreenDialog: action.fullscreenDialog,
      }
    case 'SET_DESKTOP_DIALOG':
      return {
        ...state,
        desktopDialog: action.desktopDialog,
      }
    case 'SET_DELETE_ALERT':
      return {
        ...state,
        deleteAlert: action.deleteAlert,
      }
    case 'SET_MENU_OPENED':
      return {
        ...state,
        menuOpened: action.menuOpened,
      }
    case 'SET_LABEL_EDITTING':
      return {
        ...state,
        labelEditting: action.labelEditting,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    case 'ADD_LABEL':
      return {
        ...state,
        data: [...state.data, action.label],
        labelSelected: state.labelSelected || action.label,
      }
    case 'EDIT_LABEL':
      const updatedLabels = state.data.map((label) => {
        if (label.id === action.id) {
          return {
            ...label,
            ...action.updates,
          }
        } else {
          return label
        }
      })
      return {
        ...state,
        data: updatedLabels,
        labelSelected:
          state.labelSelected && state.labelSelected.id === action.id
            ? { ...state.labelSelected, ...action.updates }
            : state.labelSelected,
      }
    case 'DELETE_LABEL':
      return {
        ...state,
        data: state.data.filter(({ id }) => id !== action.id),
        labelSelected:
          state.labelSelected && state.labelSelected.id === action.id
            ? null
            : state.labelSelected,
      }
    case 'SET_LABELS':
      return {
        ...state,
        data: action.labels,
      }
    case 'SET_LABEL_SELECTED':
      return {
        ...state,
        labelSelected: action.labelSelected,
      }
    default:
      return state
  }
}
