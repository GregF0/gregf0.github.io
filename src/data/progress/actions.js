export const setProgressVisibility = (visibility) => ({
  type: 'SET_PROGRESS_VISIBILITY',
  visibility,
})
