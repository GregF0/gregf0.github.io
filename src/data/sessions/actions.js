import fs from '../../firebase/firebase'

export const setSessions = (sessions) => ({
  type: 'SET_SESSIONS',
  sessions,
})

export const startSetSessions = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const sessionsRef = await fs.collection(`users/${uid}/sessions`).get()

      const sessions = sessionsRef.docs.map((session) => ({
        id: session.id,
        ...session.data(),
      }))

      dispatch(setSessions(sessions))
      return sessionsRef
    } catch (e) {
      dispatch(setSessions([]))
    }
  }
}

export const addSession = (session) => ({
  type: 'ADD_SESSION',
  session,
})

export const startAddSession = (session) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid
    const { duration, label = null, createdAt } = session

    const newSessionRef = fs.collection(`users/${uid}/sessions`).doc()

    dispatch(
      addSession({
        id: newSessionRef.id,
        ...session,
      })
    )

    await newSessionRef.set({
      duration,
      label,
      createdAt,
    })

    return newSessionRef
  }
}
