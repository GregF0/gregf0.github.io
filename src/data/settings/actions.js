import fs from '../../firebase/firebase'

export const setWorkDuration = (duration) => ({
  type: 'SET_WORK_DURATION',
  duration,
})

export const startSetWorkDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setWorkDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { workDuration: duration } }, { merge: true })
  }
}

export const setShortBreakDuration = (duration) => ({
  type: 'SET_SHORT_BREAK_DURATION',
  duration,
})

export const startSetShortBreakDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShortBreakDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { shortBreakDuration: duration } }, { merge: true })
  }
}

export const setLongBreakDuration = (duration) => ({
  type: 'SET_LONG_BREAK_DURATION',
  duration,
})

export const startSetLongBreakDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setLongBreakDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { longBreakDuration: duration } }, { merge: true })
  }
}

export const setRounds = (rounds) => ({
  type: 'SET_ROUNDS',
  rounds,
})

export const startSetRounds = (rounds) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setRounds(rounds))

    await fs.doc(`users/${uid}`).set({ settings: { rounds } }, { merge: true })
  }
}

export const setShowTimerInTitle = (showTimerInTitle) => ({
  type: 'SET_SHOW_TIMER_IN_TITLE',
  showTimerInTitle,
})

export const startSetShowTimerInTitle = (showTimerInTitle) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShowTimerInTitle(showTimerInTitle))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { showTimerInTitle } }, { merge: true })
  }
}

export const setShowNotifications = (showNotifications) => ({
  type: 'SET_SHOW_NOTIFICATIONS',
  showNotifications,
})

export const startSetShowNotifications = (showNotifications) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShowNotifications(showNotifications))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { showNotifications } }, { merge: true })
  }
}

export const setDarkMode = (darkMode) => ({
  type: 'SET_DARK_MODE',
  darkMode,
})

export const startSetDarkMode = (darkMode) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    localStorage.setItem('darkMode', +darkMode)

    dispatch(setDarkMode(darkMode))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { darkMode } }, { merge: true })
  }
}

export const setAutostart = (autostart) => ({
  type: 'SET_AUTOSTART',
  autostart,
})

export const startSetAutostart = (autostart) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setAutostart(autostart))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { autostart } }, { merge: true })
  }
}

export const setFirstDayOfTheWeek = (firstDayOfTheWeek) => ({
  type: 'SET_FIRST_DAY_OF_THE_WEEK',
  firstDayOfTheWeek,
})

export const startSetFirstDayOfTheWeek = (firstDayOfTheWeek) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setFirstDayOfTheWeek(firstDayOfTheWeek))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { firstDayOfTheWeek } }, { merge: true })
  }
}

export const setSettings = (settings) => ({
  type: 'SET_SETTINGS',
  settings,
})

export const startSetSettings = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const docRef = await fs.doc(`users/${uid}`).get()
      let data = {}

      if (docRef.exists) {
        data = docRef.data().settings
      }

      dispatch(setSettings(data))
    } catch (e) {
      dispatch(setSettings({}))
    }
  }
}
