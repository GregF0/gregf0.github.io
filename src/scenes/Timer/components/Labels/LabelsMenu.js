import React, { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import styled from 'styled-components'
import AddIcon from '@material-ui/icons/Add'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import { useTheme, useMediaQuery } from '@material-ui/core'
import { Label } from './Label'
import {
  setFullscreenDialog,
  setDesktopDialog,
  setMenuOpened,
  setLabelSelected,
} from '../../../../data/labels/actions'

export const LabelsMenu = ({ anchor }) => {
  const [anchorEl, setAnchorEl] = useState(anchor || null)

  const { data, labelSelected, menuOpened } = useSelector(
    (state) => state.labels
  )
  const dispatch = useDispatch()

  const theme = useTheme()
  const isMediumScreen = useMediaQuery(theme.breakpoints.up('md'))

  useEffect(() => {
    if (anchor) {
      setAnchorEl(anchor)
    }
  }, [anchor])

  const handleClose = (label) => {
    dispatch(setMenuOpened(false))

    if (label && label.id && label.name && label.color) {
      dispatch(setLabelSelected(label))
    }
  }

  const handleAdd = (e) => {
    e.stopPropagation()

    if (isMediumScreen) {
      dispatch(setDesktopDialog(true))
    } else {
      dispatch(setFullscreenDialog(true))
    }
  }

  return (
    <Menu
      id="long-menu"
      anchorEl={anchorEl}
      keepMounted
      open={menuOpened}
      onClose={handleClose}
      margin="dense"
    >
      {data.map((label) => (
        <LabelMenuItem
          key={label.id}
          selected={labelSelected && label.id === labelSelected.id}
          onClick={() => handleClose(label)}
          px={0}
          touchscreen={'ontouchstart' in document.documentElement ? 1 : 0}
        >
          <Label label={label} />
        </LabelMenuItem>
      ))}
      <MenuItem onClick={handleAdd}>
        <AddLabelIcon />
        Add new label
      </MenuItem>
    </Menu>
  )
}

const LabelMenuItem = styled(MenuItem)`
  .MuiIconButton-root {
    visibility: ${({ touchscreen }) => (touchscreen ? 'unset' : 'hidden')};
  }

  &:hover .MuiIconButton-root {
    visibility: ${({ touchscreen }) => (touchscreen ? 'unset' : 'visible')};
  }
`

const AddLabelIcon = styled(AddIcon)`
  width: 20px;
  height: 20px;
  margin-right: 20px;
`

export { LabelMenuItem }
