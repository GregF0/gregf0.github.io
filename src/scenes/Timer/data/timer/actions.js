export const setStatus = (status) => ({
  type: 'SET_STATUS',
  status,
})

export const setProgress = (progress) => ({
  type: 'SET_PROGRESS',
  progress,
})

export const setTimeLeft = (timeLeft) => ({
  type: 'SET_TIME_LEFT',
  timeLeft,
})

export const saveInterval = (interval) => ({
  type: 'SAVE_INTERVAL',
  interval,
})

export const pauseTimer = () => ({
  type: 'PAUSE_TIMER',
})

export const resetTimer = ({ duration, showTimerInTitle }) => ({
  type: 'RESET_TIMER',
  duration,
  showTimerInTitle,
})

export const setNextTimer = (settings) => ({
  type: 'SET_NEXT_TIMER',
  settings,
})

export const setSaveSessionAlert = (saveSessionAlert) => ({
  type: 'SET_SAVE_SESSION_ALERT',
  saveSessionAlert,
})
