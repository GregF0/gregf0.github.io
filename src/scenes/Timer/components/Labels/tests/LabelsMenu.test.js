import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import MenuItem from '@material-ui/core/MenuItem'
import labels from '../../../../../data/labels/tests/mock-data/labels'
import { LabelsMenu, LabelMenuItem } from '../LabelsMenu'
import * as labelsActions from '../../../../../data/labels/actions'

describe('<LabelsMenu />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<LabelsMenu anchor={document.createElement('div')} />)
  }

  const dispatchMocked = jest.fn()
  jest.spyOn(redux, 'useDispatch').mockImplementation(() => dispatchMocked)

  const createStore = (labelSelected = null) => {
    const store = {
      labels: {
        labelSelected,
        data: labels,
        menuOpened: true,
      },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <LabelsMenu /> correctly', () => {
    createStore()
    expect(createWrapper()).toMatchSnapshot()
  })

  test('only third menu item should be selected', () => {
    createStore(labels[2])
    const wrapper = createWrapper()

    expect(wrapper.find(LabelMenuItem).at(0).prop('selected')).toBe(false)

    expect(wrapper.find(LabelMenuItem).at(1).prop('selected')).toBe(false)

    expect(wrapper.find(LabelMenuItem).at(2).prop('selected')).toBe(true)
  })

  test('should render correct number of menu items', () => {
    createStore()
    expect(createWrapper().find(LabelMenuItem).length).toBe(labels.length)
  })

  test('should handle add button', () => {
    const setFullscreenDialogMocked = jest
      .spyOn(labelsActions, 'setFullscreenDialog')
      .mockImplementation(() => jest.fn())

    createStore()
    createWrapper()
      .find(MenuItem)
      .simulate('click', { stopPropagation: jest.fn() })
    expect(setFullscreenDialogMocked).toHaveBeenCalledWith(true)
  })
})
