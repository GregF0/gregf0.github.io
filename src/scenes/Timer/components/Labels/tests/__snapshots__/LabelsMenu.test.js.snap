// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LabelsMenu /> should render <LabelsMenu /> correctly 1`] = `
<WithStyles(ForwardRef(Menu))
  anchorEl={<div />}
  id="long-menu"
  keepMounted={true}
  margin="dense"
  onClose={[Function]}
  open={true}
>
  <Styled(WithStyles(ForwardRef(MenuItem)))
    key="1"
    onClick={[Function]}
    px={0}
    selected={null}
    touchscreen={0}
  >
    <Label
      label={
        Object {
          "color": "red",
          "id": "1",
          "name": "Coding",
        }
      }
    />
  </Styled(WithStyles(ForwardRef(MenuItem)))>
  <Styled(WithStyles(ForwardRef(MenuItem)))
    key="2"
    onClick={[Function]}
    px={0}
    selected={null}
    touchscreen={0}
  >
    <Label
      label={
        Object {
          "color": "blue",
          "id": "2",
          "name": "Reading",
        }
      }
    />
  </Styled(WithStyles(ForwardRef(MenuItem)))>
  <Styled(WithStyles(ForwardRef(MenuItem)))
    key="3"
    onClick={[Function]}
    px={0}
    selected={null}
    touchscreen={0}
  >
    <Label
      label={
        Object {
          "color": "green",
          "id": "3",
          "name": "Playing a guitar",
        }
      }
    />
  </Styled(WithStyles(ForwardRef(MenuItem)))>
  <WithStyles(ForwardRef(MenuItem))
    onClick={[Function]}
  >
    <Styled(AddIcon) />
    Add new label
  </WithStyles(ForwardRef(MenuItem))>
</WithStyles(ForwardRef(Menu))>
`;
